### DISCUSSION TITLE
node await not closes

### DISCUSSION THREAD

**Discussion by rezayi** [caoccao/Javet#152] [Q&A]
Hi every body.
I have a problem with javet on 'Node' mode. I'm using javet + jsdom inside it as ssr module.
Main functionality is OK and i can render a page implemented using React.
But the problem is closing NodeRuntime!
I need to use runtime.await to activate node event-loop. As it's described in below link, I've created a worker thread and called 'runtime.await' inside it. But unfortunately i can not close the worker thread and NodeRuntime without closing whole application. Can anyone help me?
https://www.caoccao.com/Javet/tutorial/advanced/interact_with_node_js.html

**Comment by caoccao** [comment]
Could you leave a minimal reproducible code snippet? I think your JS code brings up an event loop so that Node.js runtime refuses to exit. The key is to close that event loop in JS.

**Comment by rezayi** [comment]
Thank you for answering.
these are libraries that i'm using them:
```require("fake-indexeddb/auto");
require('raf/polyfill');
require('intl-pluralrules');
Object.assign(global, {WebSocket: require('ws')});
var jsdom = require("jsdom");
```
the other scripts that i execute them are a web page scripts that implemented using react.

**Comment by caoccao** [comment]
That is not a complete code snippet. I think in your case, you don't need to call `await()` because `await()` is designed for hosting an event loop in JVM. Obviously, jsdom doesn't have an event loop, so calling `await()` will actually block forever. Please let me know if it works.

**Comment by rezayi** [comment]
Sorry, but the other part of scripts that will execute are not mine. they are the scripts of the websites, that includes some http requests such as rest calls. when i don't call `await()` the rest calls does not call.

**Comment by caoccao** [comment]
Then, I suspect there are some http connections not closed in your application. Node.js is designed to drain the event loop before exiting. I suggest you use something like fiddler/wireshark to monitor the http connections to find out which ones are blocked.

**Reply by rezayi** [reply]
unfortunately , I couldn't find any thing that blocks the v8 process after close.
I finally decided to separate the process that runs the v8, from coordinator process and kill the process after running finished

**Reply by caoccao** [reply]
Thank you for the update. That's not ideal, but works (with potential memory leak).

I still would like to point out that there are highly likely some async calls (e.g. timers, promises) not resolved. Locating and terminating those async calls would be the better solution.

**Reply by caoccao** [reply]
By the way, you don't really need to kill that thread. Instead, you may [terminate](https://www.caoccao.com/Javet/reference/troubleshooting/termination.html) the script execution to get a graceful termination without any memory leak.
